File: src/csharp/Intel/Iced/Intel/DecoderInternal/TableDeserializer.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if DECODER
using System;
using System.Collections.Generic;
using Iced.Intel.Internal;

namespace Iced.Intel.DecoderInternal {
	abstract class OpCodeHandlerReader {
		public abstract int ReadHandlers(ref TableDeserializer deserializer, OpCodeHandler?[] result, int resultIndex);
	}

	readonly struct HandlerInfo {
		public readonly OpCodeHandler? handler;
		public readonly OpCodeHandler?[]? handlers;

		public HandlerInfo(OpCodeHandler handler) {
			this.handler = handler;
			handlers = null;
		}

		public HandlerInfo(OpCodeHandler?[] handlers) {
			handler = null;
			this.handlers = handlers;
		}
	}

#if HAS_SPAN
	ref struct TableDeserializer {
#else
	struct TableDeserializer {
#endif
		DataReader reader;
		readonly OpCodeHandlerReader handlerReader;
		readonly List<HandlerInfo> idToHandler;
		readonly OpCodeHandler?[] handlerArray;

#if HAS_SPAN
		public TableDeserializer(OpCodeHandlerReader handlerReader, int maxIds, ReadOnlySpan<byte> data) {
#else
		public TableDeserializer(OpCodeHandlerReader handlerReader, int maxIds, byte[] data) {
#endif
			this.handlerReader = handlerReader;
			reader = new DataReader(data);
			idToHandler = new List<HandlerInfo>(maxIds);
			handlerArray = new OpCodeHandler[1];
		}

		public void Deserialize() {
			for (; reader.CanRead;) {
				switch ((SerializedDataKind)reader.ReadByte()) {
				case SerializedDataKind.HandlerReference:
					idToHandler.Add(new HandlerInfo(ReadHandler()));
					break;

				case SerializedDataKind.ArrayReference:
					idToHandler.Add(new HandlerInfo(ReadHandlers((int)reader.ReadCompressedUInt32())));
					break;

				default:
					throw new InvalidOperationException();
				}
			}
			if (reader.CanRead)
				throw new InvalidOperationException();
		}

		public LegacyOpCodeHandlerKind ReadLegacyOpCodeHandlerKind() => (LegacyOpCodeHandlerKind)reader.ReadByte();
#if !NO_VEX || !NO_XOP
		public VexOpCodeHandlerKind ReadVexOpCodeHandlerKind() => (VexOpCodeHandlerKind)reader.ReadByte();
#endif
#if !NO_EVEX
		public EvexOpCodeHandlerKind ReadEvexOpCodeHandlerKind() => (EvexOpCodeHandlerKind)reader.ReadByte();
#endif
#if MVEX
		public MvexOpCodeHandlerKind ReadMvexOpCodeHandlerKind() => (MvexOpCodeHandlerKind)reader.ReadByte();
#endif
		public Code ReadCode() => (Code)reader.ReadCompressedUInt32();
		public Register ReadRegister() => (Register)reader.ReadByte();
		public DecoderOptions ReadDecoderOptions() => (DecoderOptions)reader.ReadCompressedUInt32();
		public HandlerFlags ReadHandlerFlags() => (HandlerFlags)reader.ReadCompressedUInt32();
		public LegacyHandlerFlags ReadLegacyHandlerFlags() => (LegacyHandlerFlags)reader.ReadCompressedUInt32();
#if !NO_EVEX
		public TupleType ReadTupleType() => (TupleType)reader.ReadByte();
#endif
		public bool ReadBoolean() => reader.ReadByte() != 0;
		public int ReadInt32() => (int)reader.ReadCompressedUInt32();

		public OpCodeHandler ReadHandler() => ReadHandlerOrNull() ?? throw new InvalidOperationException();

		public OpCodeHandler? ReadHandlerOrNull() {
			int count = handlerReader.ReadHandlers(ref this, handlerArray, 0);
			if (count != 1)
				throw new InvalidOperationException();
			return handlerArray[0];
		}

		public OpCodeHandler?[] ReadHandlers(int count) {
			var handlers = new OpCodeHandler?[count];
			for (int i = 0; i < handlers.Length;) {
				int num = handlerReader.ReadHandlers(ref this, handlers, i);
				if (num <= 0 || (uint)i + (uint)num > (uint)handlers.Length)
					throw new InvalidOperationException();
				i += num;
			}
			return handlers;
		}

		public OpCodeHandler ReadHandlerReference() {
			uint index = reader.ReadByte();
			return idToHandler[(int)index].handler ?? throw new InvalidOperationException();
		}

		public OpCodeHandler[] ReadArrayReference(uint kind) {
			if (reader.ReadByte() != kind)
				throw new InvalidOperationException();
			return GetTable(reader.ReadByte());
		}

		public OpCodeHandler[] GetTable(uint index) =>
			(idToHandler[(int)index].handlers ?? throw new InvalidOperationException())!;
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/dec/TableDeserializer.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.dec;

import java.util.ArrayList;

import com.github.icedland.iced.x86.internal.dec.SerializedDataKind;

abstract class OpCodeHandlerReader {
	abstract int readHandlers(TableDeserializer deserializer, OpCodeHandler[] result, int resultIndex);
}

final class HandlerInfo {
	final OpCodeHandler handler;
	final OpCodeHandler[] handlers;

	HandlerInfo(OpCodeHandler handler) {
		this.handler = handler;
		handlers = null;
	}

	HandlerInfo(OpCodeHandler[] handlers) {
		handler = null;
		this.handlers = handlers;
	}
}

@SuppressWarnings("deprecation")
final class TableDeserializer {
	private final com.github.icedland.iced.x86.internal.DataReader reader;
	private final OpCodeHandlerReader handlerReader;
	private final ArrayList<HandlerInfo> idToHandler;
	private final OpCodeHandler[] handlerArray;

	TableDeserializer(OpCodeHandlerReader handlerReader, int maxIds, byte[] data) {
		this.handlerReader = handlerReader;
		reader = new com.github.icedland.iced.x86.internal.DataReader(data);
		idToHandler = new ArrayList<HandlerInfo>(maxIds);
		handlerArray = new OpCodeHandler[1];
	}

	void deserialize() {
		for (; reader.canRead();) {
			switch (reader.readByte()) {
			case SerializedDataKind.HANDLER_REFERENCE:
				idToHandler.add(new HandlerInfo(readHandler()));
				break;

			case SerializedDataKind.ARRAY_REFERENCE:
				idToHandler.add(new HandlerInfo(readHandlers(reader.readCompressedUInt32())));
				break;

			default:
				throw new UnsupportedOperationException();
			}
		}
		if (reader.canRead())
			throw new UnsupportedOperationException();
	}

	int readLegacyOpCodeHandlerKind() {
		return reader.readByte();
	}

	int readVexOpCodeHandlerKind() {
		return reader.readByte();
	}

	int readEvexOpCodeHandlerKind() {
		return reader.readByte();
	}

	int readMvexOpCodeHandlerKind() {
		return reader.readByte();
	}

	int readCode() {
		return reader.readCompressedUInt32();
	}

	int readRegister() {
		return reader.readByte();
	}

	int readDecoderOptions() {
		return reader.readCompressedUInt32();
	}

	int readHandlerFlags() {
		return reader.readCompressedUInt32();
	}

	int readLegacyHandlerFlags() {
		return reader.readCompressedUInt32();
	}

	int readTupleType() {
		return reader.readByte();
	}

	boolean readBoolean() {
		return reader.readByte() != 0;
	}

	int readInt32() {
		return reader.readCompressedUInt32();
	}

	OpCodeHandler readHandler() {
		OpCodeHandler handler = readHandlerOrNull();
		if (handler == null)
			throw new UnsupportedOperationException();
		return handler;
	}

	OpCodeHandler readHandlerOrNull() {
		int count = handlerReader.readHandlers(this, handlerArray, 0);
		if (count != 1)
			throw new UnsupportedOperationException();
		return handlerArray[0];
	}

	OpCodeHandler[] readHandlers(int count) {
		OpCodeHandler[] handlers = new OpCodeHandler[count];
		for (int i = 0; i < handlers.length;) {
			int num = handlerReader.readHandlers(this, handlers, i);
			if (num <= 0 || Integer.compareUnsigned(i + num, handlers.length) > 0)
				throw new UnsupportedOperationException();
			i += num;
		}
		return handlers;
	}

	OpCodeHandler readHandlerReference() {
		int index = reader.readByte();
		OpCodeHandler handler = idToHandler.get(index).handler;
		if (handler == null)
			throw new UnsupportedOperationException();
		return handler;
	}

	OpCodeHandler[] readArrayReference(int kind) {
		if (reader.readByte() != kind)
			throw new UnsupportedOperationException();
		return getTable(reader.readByte());
	}

	OpCodeHandler[] getTable(int index) {
		OpCodeHandler[] handlers = idToHandler.get(index).handlers;
		if (handlers == null)
			throw new UnsupportedOperationException();
		return handlers;
	}
}
